[FIX][14.0][vault] Lock out of users when login changed

fkantelberg opened the pull request:
During the work on disaster recovery I discovered a bug regarding the decryption of the private key. Previously the password for encrypting/decrypting the private key was formed from the login of the user and the given password and/or passfile. If the user logjn is changed the first part and therefore the password will always be wrong locking out the user from his private key and all vaults effectivly.

The fix is taking care of it by introducing a versioning to the res.users.key to indicate if a migration of the key is required. The javascript should migrate the private keys with the next usage automatically. The private key will be re-encrypting without the login and stored in the database as a new res.users.key with version 1.

It will only cause a problem if the login already changed. In that case it's recommended to change the login back shortly. Alternativly the res.users.key could be flagged as version 1 and the user should input `<username>|<password>`.

fkantelberg committed:
Prevent locking out the user from the private key when the login was changed

Note about disaster recovery project

Add missing dependency to base_setup for res.config.settings view

fkantelberg commented:
@AntoniRomera Could you also please review here?

fkantelberg commented:
@rousseldenis @CarlosRoca13 Could you also please review this PR?

CarlosRoca13 commented:
When I try to create a new private key is doing nothing:
![01](https://user-images.githubusercontent.com/35952655/171106001-e0bf4464-c9ae-44ed-93e4-c43981cdf4ad.gif)

fkantelberg commented:
@CarlosRoca13 Thank you for testing.

I suspect the missing secure context behind the doing nothing. This modules requires a HTTPS connector or to run on localhost. Both are required for the secure context of the browsers which functions are used.

CarlosRoca13 commented:
When I'm trying to install it locally I'm getting an error:

`ValueError: External ID not found in the system: base_setup.res_config_settings_view_form`

fkantelberg commented:
@CarlosRoca13 I totally forgot the dependency. I pushed the fix to the branch here. Thank you

CarlosRoca13 commented on vault/__manifest__.py:
```diff
@@ -5,12 +5,12 @@
     "summary": "Password vault integration in Odoo",
     "license": "AGPL-3",
-    "version": "14.0.1.6.0",
+    "version": "14.0.1.6.1",
```
This change will be done by the bot, please restore to the previous version

fkantelberg replied:
Done

CarlosRoca13 reviewed (commented):
Just a comment

It is working well

pedrobaeza commented:
Please squash commit history according the business logic separation you want to do

fkantelberg commented:
@pedrobaeza I squashed everything

pedrobaeza commented:
What is going to happen with existing module users? Is this compatible?

fkantelberg commented:
@pedrobaeza 
Most of the stuff I added was to ensure it's backward compatible. I added a version field to the model storing the keys in the database and a on-the-fly migration in the javascript. Because it requires a decrypted private key it's only possible when a user decrypts it. In case the private key is loaded with the old version it's re-encrypted with the new material and will create a new entry in the database. This is done via the javascript `_check_key_migration`.

New keys will always be created with the current version 1. If we need to change something later again we can do the migration the same way. This makes it basically backwards compatible without breaking.

You are right asking because it's crucial. This was also the reason I created and added the link to vault-recovery to support a disaster recovery from backups more easily.

pedrobaeza commented:
Great, thanks for the detailed explanation.

There's an error in the CI that I think is not related to this PR. I'll switch as part of my community involvement to GH actions and see if the CI becomes green.

fkantelberg commented:
@pedrobaeza The error is in the module `auth_signup_verify_email` in one of the test cases. Quite strange but I see no relation here. Do you need anything else from me here?

pedrobaeza commented:
/ocabot merge patch
